promremote/client.go

```diff
@@ -0,0 +1,178 @@
+	Value string `json:"value"`
+}
+
+// Timeseries are made of tags and a datapoint.
+type Timeseries struct {
+	Tags      []Tag
+	Datapoint ts.Datapoint
+}
+
+type client struct {
+	writeURL   string
+	httpClient *http.Client
+}
+
+type clientOptions struct {
+	remoteWriteURL    string
+	httpClientTimeout time.Duration
+}
+
+// M3Client is used to write timeseries data to m3coordinator.
+type M3Client interface {
+	// Write writes the Prom proto WriteRequest to the specified endpoint.
+	Write(*prompb.WriteRequest) error
+}
+
+// ClientOptions defines available methods.
+type ClientOptions interface {
+	// SetWriteURL sets the URL which the client uses to write to m3coordinator.
+	SetWriteURL(string) ClientOptions
+
+	// WriteURL returns the URL which the client uses to write to m3coordinator.
+	WriteURL() string
+
+	// SetHTTPClientTimeout sets the timeout for the client.
+	SetHTTPClientTimeout(time.Duration) ClientOptions
+
+	//HTTPClientTimeout returns the timeout that is set for the client.
+	HTTPClientTimeout() time.Duration
+}
+
+// NewClientOpts returns a default clientOptions struct.
+func NewClientOpts() ClientOptions {
+	return &clientOptions{
+		remoteWriteURL:    DefaultRemoteWrite,
+		httpClientTimeout: defaulHTTPClientTimeout,
+	}
+}
+
+// NewClient creates a new remote write coordinator client.
+func NewClient(opts ClientOptions) M3Client {
+	return &client{
+		writeURL: opts.WriteURL(),
+		httpClient: &http.Client{
+			Timeout: opts.HTTPClientTimeout(),
+		},
+	}
+}
+
+func (o *clientOptions) SetWriteURL(val string) ClientOptions {
+	opts := *o
+	opts.remoteWriteURL = val
+	return &opts
+}
+
+func (o *clientOptions) WriteURL() string {
+	return o.remoteWriteURL
+}
+
+func (o *clientOptions) SetHTTPClientTimeout(val time.Duration) ClientOptions {
+	opts := *o
+	opts.httpClientTimeout = val
+	return &opts
+}
+
+func (o *clientOptions) HTTPClientTimeout() time.Duration {
+	return o.httpClientTimeout
+}
+
+func (c *client) Write(promWR *prompb.WriteRequest) error {
+	data, err := proto.Marshal(promWR)
+	if err != nil {
+		return fmt.Errorf("unable to marshal protobuf: %v", err)
+	}
+
+	encoded := snappy.Encode(nil, data)
+
+	body := bytes.NewReader(encoded)
+	req, err := http.NewRequest("POST", c.writeURL, body)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("Content-Encoding", "snappy")
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("unable to read response body: %v", err)
+		}
+
+		body := string(bodyBytes)
+		return fmt.Errorf("expected 200 response code, instead got: %d, %s", resp.StatusCode, body)
+	}
+
+	return nil
+}
+
+// TSListToProtoWR converts a list of timeseries to a Prometheus proto write request.
+func TSListToProtoWR(tsList []Timeseries) *prompb.WriteRequest {
```

Comment:
nit: maybe `TSListToWriteRequest`? alternatively, maybe add `ToPromWriteRequest` as a member function to `type TSList []Timeseries` and use that type instead?

Reply:
Or you could also just add two methods to the client, i.e.:

```go
WriteProto(req *prompb.WriteRequest) error
WriteTimeSeries(seriesList []TimeSeries) error
```

Reply:
Then I would also remove the method name.

Reply:
P.s. my suggestion does not stop you also doing Matt's suggestion to making the timeseries have a member function to return the prom request.

i.e.

```go
type TimeSeriesList []TimeSeries

func (tsList TimeSeriesList) ToWriteRequest() *prompb.WriteRequest {
	promTS := make([]*prompb.TimeSeries, len(tsList))

 	for i, ts := range tsList {
		labels := make([]*prompb.Label, len(ts.Tags))
		for j, tag := range ts.Tags {
			labels[j] = &prompb.Label{Name: tag.Name, Value: tag.Value}
		}

 		sample := []prompb.Sample{prompb.Sample{Value: ts.Datapoint.Value, Timestamp: ts.Datapoint.Timestamp.Unix()}}
		promTS[i] = &prompb.TimeSeries{Labels: labels, Samples: sample}
	}

 	return &prompb.WriteRequest{
		Timeseries: promTS,
	}
}

```

Then making the `WriteTimeSeries` method use it:

```go
func (c *client) WriteTimeSeries(seriesList []TimeSeries) error {
  return c.WriteProto(seriesList.ToWriteRequest())
}
```